swig/cpp/src/Tree.cpp

```diff
@@ -0,0 +1,299 @@
+/**
+ * @file Trees.cpp
+ * @author Mislav Novakovic <mislav.novakovic@sartura.hr>
+ * @brief Sysrepo class header implementation for C header trees.h
+ *
+ * @copyright
+ * Copyright 2016 Deutsche Telekom AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Sysrepo.h"
+#include "Struct.h"
+#include "Tree.h"
+
+using namespace std;
+
+extern "C" {
+#include "sysrepo.h"
+#include "sysrepo/trees.h"
+}
+
+Tree::Tree(const char *root_name, const char *root_module_name) {
+    sr_node_t *node;
+    int ret = sr_new_tree(root_name, root_module_name, &node);
+    if (ret != SR_ERR_OK)
+        throw_exception(ret);
+
+    _free = true;
+    _node = node;
+}
+Tree::Tree(sr_node_t *tree, bool free) {_node = tree; _free = free;}
+Tree::~Tree() {
+    if (_node != NULL && _free)
+        sr_free_tree(_node);
```

Comment:
I too was thinking of this. I wanted to know the relation between tree and node before I implemented it. I will use std::enable_shared_from_this to get the shared_ptr of the parent and pass it to children.